app/app.go

```diff
@@ -0,0 +1,199 @@
+package app
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"io"
+	"strings"
+
+	"github.com/datatrails/go-datatrails-logverification/logverification/app"
+)
+
+/**
+ * Merklelog related APP contents
+ *
+ * An APP in the context of merklelog is an interface that commits log entries.
+ *
+ * Apps include:
+ *  - assetsv2
+ *  - eventsv1
+ */
+
+const (
+	AssetsV2AppDomain = byte(0)
+	EventsV1AppDomain = byte(1)
+)
+
+var (
+	ErrNoJsonGiven = errors.New("no json given")
+)
+
+// AppDataToVerifiableLogEntries converts the app data to verifiable log entries
+func AppDataToVerifiableLogEntries(appData []byte, logTenant string) ([]app.AppEntry, error) {
+
+	// first attempt to convert the appdata to a list of events
+	eventList, err := eventListFromData(appData)
+	if err != nil {
+		return nil, err
+	}
+
+	// now we have an event list we can decipher if the app is
+	//  assetsv2 or eventsv1
+	appDomain := appDomain(appData)
+
+	verifiableLogEntries := []app.AppEntry{}
+
+	switch appDomain {
+	case AssetsV2AppDomain:
+		// assetsv2
+		verfiableAssetsV2Events, err := NewAssetsV2AppEntries(eventList)
+		if err != nil {
+			return nil, err
+		}
+
+		verifiableLogEntries = append(verifiableLogEntries, verfiableAssetsV2Events...)
+
+	case EventsV1AppDomain:
+		verfiableEventsV1Events, err := NewEventsV1AppEntries(eventList, logTenant)
+		if err != nil {
+			return nil, err
+		}
+
+		verifiableLogEntries = append(verifiableLogEntries, verfiableEventsV1Events...)
+
+	default:
+		return nil, errors.New("unknown app domain for given app data")
+	}
+
+	return verifiableLogEntries, nil
+}
+
+// appDomain returns the app domain of the given app data
+func appDomain(appData []byte) byte {
+
+	// first attempt to convert the appdata to a list of events
+	eventList, err := eventListFromData(appData)
+	if err != nil {
+		// if we can't return default of assetsv2
+		return AssetsV2AppDomain
+	}
+
+	// decode into events
+	events := struct {
+		Events        []json.RawMessage `json:"events,omitempty"`
+		NextPageToken json.RawMessage   `json:"next_page_token,omitempty"`
+	}{}
+
+	decoder := json.NewDecoder(bytes.NewReader(eventList))
+	decoder.DisallowUnknownFields()
+	for {
+		err = decoder.Decode(&events)
+
+		if errors.Is(err, io.EOF) {
+			break
+		}
+
+		if err != nil {
+			// return default of assetsv2
+			return AssetsV2AppDomain
+		}
+	}
+
+	// decode the first event and find the identity
+	event := struct {
+		Identity string `json:"identity,omitempty"`
+	}{}
+
+	decoder = json.NewDecoder(bytes.NewReader(events.Events[0]))
+	decoder.DisallowUnknownFields()
+
+	for {
+		err = decoder.Decode(&event)
```

Comment:
I assume this is decoding in chunks, hence the loop?

Reply:
yeh the loop exists because `Decode` can decode many json objects into many structs as per Paul's comment above.

So we need to ensure we loop until the serialized json is finished